### PR TITLE
Improve exponent function

### PR DESCRIPTION
**Pull request by musm**
Remove unnecessary branch/better perf

```julia
julia> @code_llvm my_exponent(-10.1)

; Function Attrs: uwtable
define i64 @julia__exponent_72099(double) #0 {
top:
  %1 = bitcast double %0 to i64
  %2 = and i64 %1, 9223372036854775807
  %3 = icmp ult i64 %2, 9218868437227405312
  br i1 %3, label %pass, label %if

L2:                                               ; preds = %pass.8, %pass
  %k.0 = phi i64 [ %5, %pass ], [ %9, %pass.8 ]
  %4 = add nsw i64 %k.0, -1023
  ret i64 %4

if:                                               ; preds = %top
  call void @jl_throw(%jl_value_t* inttoptr (i64 2153013112 to %jl_value_t*))
  unreachable

pass:                                             ; preds = %top
  %5 = lshr i64 %2, 52
  %6 = icmp ugt i64 %2, 4503599627370495
  br i1 %6, label %L2, label %if3

if3:                                              ; preds = %pass
  %7 = icmp eq i64 %2, 0
  br i1 %7, label %if4, label %pass.8

if4:                                              ; preds = %if3
  call void @jl_throw(%jl_value_t* inttoptr (i64 2153013112 to %jl_value_t*))
  unreachable

pass.8:                                           ; preds = %if3
  %8 = call i64 @llvm.ctlz.i64(i64 %2, i1 false)
  %9 = sub nsw i64 12, %8
  br label %L2
}

julia> @code_llvm exponent(-10.1)



; Function Attrs: uwtable
define i64 @julia_exponent_72000(double) #0 {
pass:
  %1 = bitcast double %0 to i64
  %2 = and i64 %1, 9218868437227405312
  %3 = icmp eq i64 %2, 0
  br i1 %3, label %if, label %L1

L1:                                               ; preds = %pass
  %4 = lshr exact i64 %2, 52
  %5 = icmp eq i64 %2, 9218868437227405312
  br i1 %5, label %if6, label %L2

L2:                                               ; preds = %L1, %pass.5
  %k.0 = phi i64 [ %4, %L1 ], [ %10, %pass.5 ]
  %6 = add nsw i64 %k.0, -1023
  ret i64 %6

if:                                               ; preds = %pass
  %7 = fcmp une double %0, 0.000000e+00
  br i1 %7, label %pass.5, label %if3

if3:                                              ; preds = %if
  call void @jl_throw(%jl_value_t* inttoptr (i64 2153013112 to %jl_value_t*))
  unreachable

pass.5:                                           ; preds = %if
  %8 = and i64 %1, 4503599627370495
  %9 = call i64 @llvm.ctlz.i64(i64 %8, i1 false)
  %10 = sub nsw i64 12, %9
  br label %L2

if6:                                              ; preds = %L1
  call void @jl_throw(%jl_value_t* inttoptr (i64 2153013112 to %jl_value_t*))
  unreachable
```

### CODE REVIEW

**Comment by musm** [comment]
What about changing this function to behave as follows (i.e. like ilog2, which is standard)

and thus things like 
```
ldexp(significand(Inf), ilog2(Inf))
```
actually work

```
    ilog2(x)

Returns the integral part of the logarithm of `abs(x)`, using base 2 for the
logarithm. In other words, this computes the binary exponent of `x` such that
``x = significand \\times 2^exponent,`` where ``significand∈\\in [1, 2)``.

Exceptional cases (where `Int` is the machine wordsize)

* `x = 0`    returns `typemin(Int)`
* `x = ±Inf` returns `typemax(Int)`
* `x = NaN`  returns `typemax(Int)`
```

**Comment by oscardssmith** [comment]
the behavior for inf and nan seem dangerous to me, couldn't that cause all sorts of corruption?

**Comment by musm** [comment]
Which are you referring to?

**Comment by oscardssmith** [comment]
x = ±Inf returns typemax(Int)
x = NaN returns typemax(Int)
These could hide buggy code and cause all sort of problems.

**Comment by musm** [comment]
Perhaps this should be a separate issue as it is independent of the PR, which just improves the current behavior.

File: base/math.jl [hunk 1]
```diff
@@ -358,18 +358,15 @@ ldexp(x::Float32,e::Integer) = ccall((:scalbnf,libm), Float32, (Float32,Int32),
...
-        x == 0 && throw(DomainError())
-        xu &= significand_mask(T)
-        m = leading_zeros(xu)-exponent_bits(T)
-        k = 1-m
-    elseif xe == exponent_mask(T) # NaN or Inf
-        throw(DomainError())
+    xs = reinterpret(Unsigned, x) & ~sign_mask(T)
+    xs >= exponent_mask(T) && return throw(DomainError()) # NaN or Inf
+    k = Int(xs >> significand_bits(T))
+    if xs <= (~exponent_mask(T) & ~sign_mask(T)) # x is subnormal
```
Comment (musm) [inline base/math.jl]: yea here and for  frexp is doesn't matter, I chose consistency with the significand function.  See https://github.com/JuliaLang/julia/pull/19492
If you prefer I can change.

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks("scalar", vs = ":master")`

**Comment by musm** [comment]
thanks!

**Pull request merged**